File: summer/WebContent/js/system/windows/controls/RowDefinition.js
/**
 * RowDefinition
 */

define(["dojo/_base/declare", "system/Type", "controls/DefinitionBase", "windows/GridLength",
        "windows/DependencyProperty", "windows/FrameworkPropertyMetadata", "windows/PropertyChangedCallback",
        "windows/ValidateValueCallback", "windows/GridUnitType"], 
		function(declare, Type, DefinitionBase, GridLength,
				DependencyProperty, FrameworkPropertyMetadata, PropertyChangedCallback,
				ValidateValueCallback, GridUnitType){
	var RowDefinition = declare("RowDefinition", DefinitionBase,{
		constructor:function(){
		}
	});
	
	Object.defineProperties(RowDefinition.prototype,{
//		public GridLength 
		Height:
		{
			get:function()
			{
				return this.UserSizeValueCache;
			},
			set:function(value)
			{
				this.SetValue(RowDefinition.HeightProperty, value);
			}
		},
//		public double 
		MinHeight:
		{
			get:function()
			{
				return this.UserMinSizeValueCache;
			},
			set:function(value)
			{
				this.SetValue(RowDefinition.MinHeightProperty, value);
			}
		},
//		public double 
		MaxHeight:
		{
			get:function()
			{
				return this.UserMaxSizeValueCache;
			},
			set:function(value)
			{
				this.SetValue(RowDefinition.MaxHeightProperty, value);
			}
		},
//		public double 
		ActualHeight:
		{
			get:function()
			{
				var result = 0.0;
				if (this.InParentLogicalTree)
				{
					result = this.Parent.GetFinalRowDefinitionHeight(this.Index);
				}
				return result;
			}
		},
//		public double 
		Offset:
		{
			get:function()
			{
				var result = 0.0;
				if (this.Index != 0)
				{
					result = this.FinalOffset;
				}
				return result;
			}
		}
	});
	
	Object.defineProperties(RowDefinition,{
//		public static readonly DependencyProperty 
		HeightProperty:
        {
        	get:function(){
        		if(RowDefinition._HeightProperty === undefined){
        			RowDefinition._HeightProperty = DependencyProperty.Register("Height", 
        	        		GridLength.Type, 
        	        		RowDefinition.Type, 
        	        		/*new FrameworkPropertyMetadata(new GridLength(1.0, GridUnitType.Star), 
        	        				new PropertyChangedCallback(null, DefinitionBase.OnUserSizePropertyChanged))*/
        	        		FrameworkPropertyMetadata.BuildWithDVandPCCB(new GridLength(1.0, GridUnitType.Star), 
        	        				new PropertyChangedCallback(null, DefinitionBase.OnUserSizePropertyChanged)), 
        	        				new ValidateValueCallback(null, DefinitionBase.IsUserSizePropertyValueValid));
        		}
        		
        		return RowDefinition._HeightProperty;
        	}
        }, 
//		public static readonly DependencyProperty 
		MinHeightProperty:
        {
        	get:function(){
        		if(RowDefinition._MinHeightProperty === undefined){
        			RowDefinition._MinHeightProperty = DependencyProperty.Register("MinHeight", 
        	        		Number.Type, RowDefinition.Type, 
        	        		/*new FrameworkPropertyMetadata(0.0, 
        	        				new PropertyChangedCallback(null, DefinitionBase.OnUserMinSizePropertyChanged))*/
        	        		FrameworkPropertyMetadata.BuildWithDVandPCCB(0.0, 
        	        				new PropertyChangedCallback(null, DefinitionBase.OnUserMinSizePropertyChanged)), 
        	        				new ValidateValueCallback(null, DefinitionBase.IsUserMinSizePropertyValueValid));
        		}
        		
        		return RowDefinition._MinHeightProperty;
        	}
        },
//		public static readonly DependencyProperty 
		MaxHeightProperty:
        {
        	get:function(){
        		if(RowDefinition._MaxHeightProperty === undefined){
        			RowDefinition._MaxHeightProperty= DependencyProperty.Register("MaxHeight", typeof(double), 
        	        		RowDefinition.Type,
        	        		/*new FrameworkPropertyMetadata(double.PositiveInfinity, 
        	        				new PropertyChangedCallback(null, DefinitionBase.OnUserMaxSizePropertyChanged))*/
        	        		FrameworkPropertyMetadata.BuildWithDVandPCCB(double.PositiveInfinity, 
        	        				new PropertyChangedCallback(null, DefinitionBase.OnUserMaxSizePropertyChanged)), 
        	        				new ValidateValueCallback(null, DefinitionBase.IsUserMaxSizePropertyValueValid));
        		}
        		
        		return RowDefinition._MaxHeightProperty;
        	}
        }, 
		  
	});
	
	RowDefinition.Type = new Type("RowDefinition", RowDefinition, [DefinitionBase.Type]);
	return RowDefinition;
});

		

File: org.summer.view.widget/src/org/summer/view/widget/controls/RowDefinition.java
/**
 */
package org.summer.view.widget.controls;

import org.summer.view.widget.DependencyObject;
import org.summer.view.widget.DependencyProperty;
import org.summer.view.widget.DependencyPropertyChangedEventArgs;
import org.summer.view.widget.FrameworkPropertyMetadata;

public class RowDefinition extends DefinitionBase {
//    #region Public Fields
//    #region Dependency Properties
//    [TypeConverter("System.Windows.LengthConverter, PresentationFramework, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, Custom=null")]
    public static final DependencyProperty MaxHeightProperty = DependencyProperty.Register("MaxHeight", typeof(Double), typeof(RowDefinition), new FrameworkPropertyMetadata(double.PositiveInfinity, InvalidateGridMeasure));
//    [TypeConverter("System.Windows.LengthConverter, PresentationFramework, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, Custom=null")]
    public static final DependencyProperty MinHeightProperty = DependencyProperty.Register("MinHeight", typeof(Double), typeof(RowDefinition), new FrameworkPropertyMetadata(InvalidateGridMeasure));
    public static final DependencyProperty HeightProperty = DependencyProperty.Register("Height", typeof(GridLength), typeof(RowDefinition), new FrameworkPropertyMetadata(new GridLength(1, GridUnitType.Star), InvalidateGridMeasure));
//    #endregion
//    #endregion
//
//    #region Private Fields
    double actual_height;
    double offset;
    Grid grid;
//    #endregion

//    #region Public Constructors
    public RowDefinition() {
    }
//    #endregion

//    #region Public Properties
//    #region Dependency Properties
//    [TypeConverter(typeof(LengthConverter))]
    public double MaxHeight {
            get { return (double)GetValue(MaxHeightProperty); }
            set { SetValue(MaxHeightProperty, value); } 
    }

//    [TypeConverter(typeof(LengthConverter))]
    public double MinHeight {
            get { return (double)GetValue(MinHeightProperty); }
            set { SetValue(MinHeightProperty, value); }
    }

    public GridLength Height {
            get { return (GridLength)GetValue(HeightProperty); }
            set { SetValue(HeightProperty, value); }
    }
//    #endregion

    public double ActualHeight {
            get { return actual_height; }
            internal set { actual_height = value; }
    }

    public double Offset {
            get { return offset; }
            internal set { offset = value; }
    }
//    #endregion

//    #region Internal Properties
    internal Grid Grid {
            get { return grid; }
            set { grid = value; }
    }
//    #endregion

//    #region Private Methods
    static void InvalidateGridMeasure(DependencyObject d, DependencyPropertyChangedEventArgs e) {
            Grid grid = ((RowDefinition)d).grid;
            if (grid != null)
                    grid.InvalidateMeasure();
    }
//    #endregion
}